File: cpp/lib/softmax.cpp
// Copyright 2026 FlagOS Contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <ATen/WrapDimUtils.h>
#include <algorithm>
#include "flag_gems/backend_utils.h"
#include "flag_gems/device_info.h"
#include "flag_gems/operators.h"
#include "flag_gems/utils.h"
#include "triton_jit/triton_jit_function.h"
namespace flag_gems {
using namespace triton_jit;

namespace {

  const TritonJITFunction &get_kernel(const std::string &name) {
    static const std::string src_path = (utils::get_flag_gems_src_path() / "ops" / "softmax.py").string();
    return TritonJITFunction::get_instance(src_path, name);
  }

  void compute_mnk(const at::Tensor &tensor, int dim, int64_t &M, int64_t &N, int64_t &K) {
    const auto sizes = tensor.sizes();
    M = 1;
    N = sizes[dim];
    K = 1;
    for (int i = 0; i < dim; ++i) M *= sizes[i];
    for (int i = dim + 1; i < sizes.size(); ++i) K *= sizes[i];
  }

  unsigned int softmax_heur_num_warps(unsigned int tile_size) {
    if (tile_size < 2048) {
      return 4;
    }
    if (tile_size < 4096) {
      return 8;
    }
    return 16;
  }

  unsigned int softmax_heur_tile_n_inner(int64_t N) {
    if (N <= 32 * 1024) {
      return static_cast<unsigned int>(utils::next_power_of_2(N));
    }
    return 4096;
  }

  unsigned int softmax_heur_tile_k(int64_t M, int64_t K) {
    constexpr int MAX_TILE_K = 8192;
    const int num_sms = std::max(1, device::current_sm_count());
    unsigned int tile_k = 1;
    const int64_t upper_bound = std::min(K, static_cast<int64_t>(MAX_TILE_K));
    while (static_cast<int64_t>(tile_k) <= upper_bound) {
      const int64_t num_blocks = M * utils::cdiv(static_cast<int>(K), static_cast<int>(tile_k));
      const double num_waves = static_cast<double>(num_blocks) / num_sms;
      if (num_waves > 1.0 && static_cast<int64_t>(tile_k) * 2 <= upper_bound) {
        tile_k *= 2;
      } else {
        break;
      }
    }
    return tile_k;
  }

  unsigned int softmax_heur_one_tile_per_cta(unsigned int tile_n, int64_t N) {
    return tile_n >= static_cast<unsigned int>(N) ? 1u : 0u;
  }

  struct SoftmaxInnerConfig {
    unsigned int tile_n;
    unsigned int one_tile_per_cta;
    unsigned int num_warps;
  };

  SoftmaxInnerConfig compute_inner_config(int64_t N) {
    const unsigned int tile_n = softmax_heur_tile_n_inner(N);
    return {tile_n, softmax_heur_one_tile_per_cta(tile_n, N), softmax_heur_num_warps(tile_n)};
  }

  struct SoftmaxNonInnerConfig {
    unsigned int tile_n;
    unsigned int tile_k;
    unsigned int one_tile_per_cta;
    unsigned int num_warps;
  };

  SoftmaxNonInnerConfig compute_forward_non_inner_config(int64_t M, int64_t N, int64_t K) {
    const unsigned int tile_k = softmax_heur_tile_k(M, K);
    const unsigned int tile_n = static_cast<unsigned int>(utils::cdiv(8192, static_cast<int>(tile_k)));
    const unsigned int tile_size = tile_n * tile_k;
    return {tile_n, tile_k, softmax_heur_one_tile_per_cta(tile_n, N), softmax_heur_num_warps(tile_size)};
  }

  SoftmaxNonInnerConfig compute_backward_non_inner_config(int64_t M, int64_t N, int64_t K) {
    const unsigned int tile_k = softmax_heur_tile_k(M, K);
    const unsigned int tile_n = std::max(1u, 1024u / tile_k);
    const unsigned int tile_size = tile_n * tile_k;
    return {tile_n, tile_k, softmax_heur_one_tile_per_cta(tile_n, N), softmax_heur_num_warps(tile_size)};
  }

  struct SoftmaxBackwardInnerConfig {
    unsigned int tile_m;
    unsigned int tile_n;
    unsigned int one_tile_per_cta;
    unsigned int num_warps;
  };

  SoftmaxBackwardInnerConfig compute_backward_inner_config(int64_t N) {
    const unsigned int tile_n = softmax_heur_tile_n_inner(N);
    const unsigned int tile_m = std::max(1u, 1024u / tile_n);
    return {tile_m, tile_n, softmax_heur_one_tile_per_cta(tile_n, N), softmax_heur_num_warps(tile_n)};
  }

  at::Tensor softmax_forward(const at::Tensor &input, int dim) {
    at::Tensor output = at::empty_like(input, input.options());

    int64_t M, N, K;
    compute_mnk(input, dim, M, N, K);

    c10::DeviceGuard guard(input.device());
    backend::StreamType stream = backend::getCurrentStream();
    backend::RawStreamType raw_stream = backend::getRawStream(stream);

    constexpr unsigned int NUM_STAGES = 1;

    if (K == 1) {
      const SoftmaxInnerConfig config = compute_inner_config(N);
      const TritonJITFunction &kernel = get_kernel("softmax_kernel_inner");
      const unsigned int grid_x = static_cast<unsigned int>(M);

      kernel(raw_stream,
             grid_x,
             1,
             1,
             config.num_warps,
             NUM_STAGES,
             output,
             input,
             M,
             N,
             config.tile_n,
             config.one_tile_per_cta);
    } else {
      const SoftmaxNonInnerConfig config = compute_forward_non_inner_config(M, N, K);
      const TritonJITFunction &kernel = get_kernel("softmax_kernel_non_inner");
      const unsigned int grid_x = static_cast<unsigned int>(M);
      const unsigned int grid_y =
          static_cast<unsigned int>(utils::cdiv(static_cast<int>(K), static_cast<int>(config.tile_k)));

      kernel(raw_stream,
             grid_x,
             grid_y,
             1,
             config.num_warps,
             NUM_STAGES,
             output,
             input,
             M,
             N,
             K,
             config.tile_n,
             config.tile_k,
             config.one_tile_per_cta);
    }

    return output;
  }

  void compute_mnk_for_backward(const at::Tensor &tensor,
                                int dim,
                                int64_t &M,
                                int64_t &N,
                                int64_t &K,
                                int64_t &stride_m,
                                int64_t &stride_n,
                                int64_t &stride_k) {
    const auto sizes = tensor.sizes();
    const auto strides = tensor.strides();

    M = 1;
    for (int i = 0; i < dim; ++i) M *= sizes[i];
    N = sizes[dim];
    K = 1;
    for (int i = dim + 1; i < sizes.size(); ++i) K *= sizes[i];

    stride_m = (dim > 0) ? strides[dim - 1] : 0;
    stride_n = strides[dim];
    stride_k = (dim + 1 < sizes.size()) ? strides[dim + 1] : 1;

    if (K == 1) stride_k = 0;
    if (M == 1) stride_m = 0;
  }

  at::Tensor softmax_backward_impl(const at::Tensor &output, const at::Tensor &grad_output, int dim) {
    at::Tensor grad_input = at::empty_like(grad_output, grad_output.options());

    int64_t M, N, K;
    int64_t stride_m, stride_n, stride_k;
    compute_mnk_for_backward(output, dim, M, N, K, stride_m, stride_n, stride_k);

    c10::DeviceGuard guard(output.device());
    backend::StreamType stream = backend::getCurrentStream();
    backend::RawStreamType raw_stream = backend::getRawStream(stream);

    constexpr unsigned int NUM_STAGES = 1;

    if (K == 1) {
      const SoftmaxBackwardInnerConfig config = compute_backward_inner_config(N);
      const TritonJITFunction &kernel = get_kernel("softmax_backward_kernel_inner");
      const unsigned int grid_x =
          static_cast<unsigned int>(utils::cdiv(static_cast<int>(M), static_cast<int>(config.tile_m)));

      kernel(raw_stream,
             grid_x,
             1,
             1,
             config.num_warps,
             NUM_STAGES,
             output,
             grad_output,
             grad_input,
             M,
             N,
             config.tile_m,
             config.tile_n,
             config.one_tile_per_cta);
    } else {
      const SoftmaxNonInnerConfig config = compute_backward_non_inner_config(M, N, K);
      const TritonJITFunction &kernel = get_kernel("softmax_backward_kernel_non_inner");
      const unsigned int grid_x = static_cast<unsigned int>(M);
      const unsigned int grid_y =
          static_cast<unsigned int>(utils::cdiv(static_cast<int>(K), static_cast<int>(config.tile_k)));

      kernel(raw_stream,
             grid_x,
             grid_y,
             1,
             config.num_warps,
             NUM_STAGES,
             output,
             grad_output,
             grad_input,
             M,
             N,
             K,
             config.tile_n,
             config.tile_k,
             config.one_tile_per_cta);
    }

    return grad_input;
  }

}  // namespace

at::Tensor softmax(const at::Tensor &input, int64_t dim, bool half_to_float) {
  int64_t dim_ = at::maybe_wrap_dim(dim, input.dim());

  const at::ScalarType out_dtype =
      (half_to_float && input.scalar_type() == at::kHalf) ? at::kFloat : input.scalar_type();

  if (input.numel() == 0) {
    at::Tensor output = at::empty_like(input, input.options().dtype(out_dtype));
    output.zero_();
    return output;
  }

  at::Tensor input_tensor = input.contiguous();
  if (half_to_float && input.scalar_type() == at::kHalf) {
    input_tensor = input_tensor.to(at::kFloat);
  }

  return softmax_forward(input_tensor, static_cast<int>(dim_));
}

at::Tensor softmax_backward(const at::Tensor &grad_output,
                            const at::Tensor &output,
                            int64_t dim,
                            at::ScalarType input_dtype) {
  int64_t wrapped_dim = at::maybe_wrap_dim(dim, output.dim());

  if (output.numel() == 0) {
    at::Tensor grad_input = at::empty_like(grad_output, grad_output.options().dtype(input_dtype));
    grad_input.zero_();
    return grad_input;
  }

  at::Tensor grad_input = softmax_backward_impl(output, grad_output, wrapped_dim);

  if (grad_input.scalar_type() != input_dtype) {
    grad_input = grad_input.to(input_dtype);
  }

  return grad_input;
}

}  // namespace flag_gems

File: src/flag_gems/runtime/backend/_kunlunxin/ops/softmax.py
# Copyright 2026 FlagOS Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

import torch
import triton
import triton.language as tl

from flag_gems import runtime
from flag_gems.ops.zeros import zero_
from flag_gems.runtime import torch_device_fn
from flag_gems.utils import libentry
from flag_gems.utils import triton_lang_extension as ext

logger = logging.getLogger(__name__)


@triton.jit
def next_multiple_of(a, b):
    # the smallest x>=a that x%b ==0
    return tl.cdiv(a, b) * b


@triton.jit
def prev_multiple_of(a, b):
    # the largest x<a that x%b ==0
    return tl.cdiv(a, b) * b - b


@libentry()
@triton.heuristics(runtime.get_heuristic_config("softmax_inner"))
@triton.jit
def softmax_kernel_inner(
    output_ptr,
    input_ptr,
    M,
    N,
    TILE_N: tl.constexpr,
    ONE_TILE_PER_CTA: tl.constexpr,
):
    pid_m = ext.program_id(0)
    if ONE_TILE_PER_CTA:
        # Pre-offset the base pointers so the inner `ptr + n_offsets` access is a
        # scalar-base + stride-1 arange that OffsetAnalysis proves contiguous
        # (block DMA). The old inline `pid_m * N + n_offsets` addressing blocked
        # the analysis -> discrete scalar gather (~1-3 GB/s, e.g. [4096,4096] took
        # ~37ms). Pre-offsetting drops it to ~1.1ms (~35x).
        input_ptr += pid_m * N
        output_ptr += pid_m * N
        n_offsets = tl.arange(0, TILE_N)
        mask = n_offsets < N
        inp = tl.load(input_ptr + n_offsets, mask=mask, other=-float("inf")).to(
            output_ptr.dtype.element_ty
        )
        m = tl.max(inp, 0)
        e = tl.exp(inp - m)
        z = tl.sum(e, 0)
        out = e / z
        tl.store(output_ptr + n_offsets, out, mask=mask)
    else:
        m = tl.full([TILE_N], value=float("-inf"), dtype=tl.float32)
        z = tl.full([TILE_N], value=0.0, dtype=tl.float32)
        input_ptr += pid_m * N
        output_ptr += pid_m * N

        previous_multiple = prev_multiple_of(N, TILE_N)
        for start_n in range(0, previous_multiple, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            inp = tl.load(input_ptr + n_offsets)
            m_new = tl.maximum(m, inp)
            # it is possible that there are -inf's in the input
            all_neg_inf = m_new == float("-inf")
            z = tl.where(all_neg_inf, z, z * tl.exp(m - m_new) + tl.exp(inp - m_new))
            m = m_new
        # specialize the last iteration
        for start_n in range(previous_multiple, N, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            mask = n_offsets < N
            inp = tl.load(input_ptr + n_offsets, mask=mask, other=-float("inf"))
            m_new = tl.maximum(m, inp)
            all_neg_inf = m_new == float("-inf")
            z = tl.where(all_neg_inf, z, z * tl.exp(m - m_new) + tl.exp(inp - m_new))
            m = m_new

        m_reduced = tl.max(m, 0)
        z = tl.sum(z * tl.exp(m - m_reduced), 0)
        m = m_reduced

        # Normalize pass. Iterate ASCENDING so each `input_ptr + n_offsets` load
        # and `output_ptr + n_offsets` store is a scalar-base + stride-1 arange
        # (block DMA). The old code walked the tiles DESCENDING
        # (`previous_multiple - start_n`) as a cache-locality trick, but on this
        # XPU the backward walk defeats OffsetAnalysis/prefetch -> discrete access
        # (~1-3 GB/s: [1024,65536] took ~154ms). Ascending drops it to ~4ms (~35x).
        previous_multiple = prev_multiple_of(N, TILE_N)
        for start_n in range(0, previous_multiple, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            inp = tl.load(input_ptr + n_offsets)
            o = tl.exp(inp - m) / z
            tl.store(output_ptr + n_offsets, o)
        for start_n in range(previous_multiple, N, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            mask = n_offsets < N
            inp = tl.load(input_ptr + n_offsets, mask=mask, other=-float("inf"))
            o = tl.exp(inp - m) / z
            tl.store(output_ptr + n_offsets, o, mask=mask)


# ------------------------  backward -------------------------------


def softmax_backward_kernel_inner_heru_tile_n(args):
    N = args["N"]
    if N <= 32768:
        return triton.next_power_of_2(N)
    return 4096


def softmax_backward_kernel_inner_heur_one_tile_per_cta(args):
    return args["TILE_N"] >= args["N"]


@libentry()
@triton.heuristics(
    values={
        "TILE_N": softmax_backward_kernel_inner_heru_tile_n,
        "ONE_TILE_PER_CTA": softmax_backward_kernel_inner_heur_one_tile_per_cta,
    },
)
@triton.jit
def softmax_backward_kernel_inner(
    out_ptr,
    out_grad_ptr,
    in_grad_ptr,
    M,
    N,
    TILE_N: tl.constexpr,
    ONE_TILE_PER_CTA: tl.constexpr,
):
    # One program per row (grid=(M,)), mirroring the forward. Pre-offset the base
    # pointers so the inner `ptr + n_offsets` access is a scalar-base + stride-1
    # arange that OffsetAnalysis proves contiguous (block DMA). The old impl used a
    # fixed grid=(12,) with a [TILE_M, TILE_N] tile whose `m_offsets[:,None]*N +
    # n_offsets` addressing blocked the analysis -> discrete scalar gather
    # (~1-3 GB/s: [4096,4096] took ~38ms). It also computed in float64 (2x traffic,
    # unnecessary). float32 accumulation matches the forward and the generic backend.
    pid_m = ext.program_id(0)
    out_ptr += pid_m * N
    out_grad_ptr += pid_m * N
    in_grad_ptr += pid_m * N
    if ONE_TILE_PER_CTA:
        n_offsets = tl.arange(0, TILE_N)
        mask = n_offsets < N
        out_tile = tl.load(out_ptr + n_offsets, mask=mask, other=0.0).to(tl.float32)
        out_grad_tile = tl.load(out_grad_ptr + n_offsets, mask=mask, other=0.0).to(
            tl.float32
        )
        scale = tl.sum(out_tile * out_grad_tile, 0)
        in_grad_tile = out_tile * (out_grad_tile - scale)
        tl.store(in_grad_ptr + n_offsets, in_grad_tile, mask=mask)
    else:
        # Pass 1: accumulate scale = sum(out * out_grad) over the row. Iterate
        # ASCENDING so each load is a scalar-base + stride-1 arange (block DMA).
        scale = tl.zeros([TILE_N], dtype=tl.float32)
        previous_multiple = prev_multiple_of(N, TILE_N)
        for start_n in range(0, previous_multiple, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            out_tile = tl.load(out_ptr + n_offsets).to(tl.float32)
            out_grad_tile = tl.load(out_grad_ptr + n_offsets).to(tl.float32)
            scale += out_tile * out_grad_tile
        for start_n in range(previous_multiple, N, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            mask = n_offsets < N
            out_tile = tl.load(out_ptr + n_offsets, mask=mask, other=0.0).to(tl.float32)
            out_grad_tile = tl.load(out_grad_ptr + n_offsets, mask=mask, other=0.0).to(
                tl.float32
            )
            scale += out_tile * out_grad_tile
        scale = tl.sum(scale, 0)  # scalar

        # Pass 2: write in_grad = out * (out_grad - scale), ASCENDING.
        for start_n in range(0, previous_multiple, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            out_tile = tl.load(out_ptr + n_offsets).to(tl.float32)
            out_grad_tile = tl.load(out_grad_ptr + n_offsets).to(tl.float32)
            in_grad_tile = out_tile * (out_grad_tile - scale)
            tl.store(in_grad_ptr + n_offsets, in_grad_tile)
        for start_n in range(previous_multiple, N, TILE_N):
            n_offsets = start_n + tl.arange(0, TILE_N)
            mask = n_offsets < N
            out_tile = tl.load(out_ptr + n_offsets, mask=mask, other=0.0).to(tl.float32)
            out_grad_tile = tl.load(out_grad_ptr + n_offsets, mask=mask, other=0.0).to(
                tl.float32
            )
            in_grad_tile = out_tile * (out_grad_tile - scale)
            tl.store(in_grad_ptr + n_offsets, in_grad_tile, mask=mask)


def softmax(self, dim, half_to_float=False):
    logger.debug("GEMS_KUNLUNXIN SOFTMAX")

    assert dim >= -self.ndim and dim < self.ndim, "Invalid dim"

    # special handling for dim = 0 and empty tensor
    if self.numel() == 0:
        out_shape = list(self.shape)
        out = torch.empty(out_shape, dtype=self.dtype, device=self.device)
        zero_(out)
        return out

    dim = dim % self.ndim
    M = 1
    N = self.shape[dim]
    for i in range(dim):
        M *= self.shape[i]  # pre_dim
    self = self.contiguous()
    if half_to_float:
        dtype = torch.float32
    else:
        dtype = self.dtype
    K = self.numel() // M // N  # post_dim

    with torch_device_fn.device(self.device):
        if K > 1:
            origin_dim = self.ndim
            if origin_dim == 3:
                m, n, k = self.shape
            else:  # 2D, dim == 0 -> M == 1
                n, k = self.shape
                m = 1
            # Rearrange [M, N, K] -> [M, K, N] so the reduced dim N is innermost
            # (the only fast axis on this XPU). Allocate the output tile directly
            # instead of `empty_like(self).view(...).transpose(...).contiguous()`,
            # which used to copy an uninitialized [M,K,N] buffer (a wasted
            # transpose-copy on top of the input transpose).
            inp_reshaped = (
                self.view(M, N, K).transpose(1, 2).contiguous().view(M * K, N)
            )
            out_reshaped = torch.empty((M * K, N), dtype=dtype, device=self.device)

            grid = lambda meta: (M * K, 1, 1)  # noqa: E731

            softmax_kernel_inner[grid](
                out_reshaped,
                inp_reshaped,
                M * K,
                N,
                buffer_size_limit=2048,
                is_use_mask_zero=True,
            )

            # Restore original layout (returns a transposed view, no copy).
            if M == 1 and origin_dim == 2:
                out = out_reshaped.view(K, N).transpose(0, 1)
            elif M == 1 and origin_dim == 3:
                out = out_reshaped.transpose(0, 1).view(m, n, k)
            else:
                out = out_reshaped.view(m, k, n).transpose(1, 2)
        else:
            out = torch.empty_like(self, dtype=dtype)
            grid = (M, 1, 1)
            softmax_kernel_inner[grid](
                out,
                self,
                M,
                N,
                buffer_size_limit=2048,
                is_use_mask_zero=True,
            )
    return out


def softmax_backward(grad_output, output, dim, input_dtype):
    logger.debug("GEMS_KUNLUNXIN SOFTMAX_VJP")

    assert dim >= -output.ndim and dim < output.ndim, "Invalid dim"
    dim = dim % output.ndim
    M = 1
    N = output.shape[dim]
    for i in range(dim):
        M *= output.shape[i]

    grad_output = grad_output.contiguous()
    output = output.contiguous()
    in_grad = torch.empty_like(output, dtype=torch.float32)
    K = output.numel() // M // N

    with torch_device_fn.device(in_grad.device):
        if K > 1:
            # how to use softmax_backward_kernel_inner?
            # some transpose and continuous
            out_grad_view = grad_output.view(M, N, K).transpose(1, 2).contiguous()
            out_view = output.view(M, N, K).transpose(1, 2).contiguous()
            # # 合并 M 和 K 维为 M' = M * K
            out_grad_reshaped = out_grad_view.view(M * K, N)
            out_reshaped = out_view.view(M * K, N)
            # 分配输入梯度的视图
            in_grad_view = in_grad.view(M, N, K).transpose(1, 2).contiguous()
            in_grad_reshaped = in_grad_view.view(M * K, N)

            grid = lambda meta: (M * K, 1, 1)  # noqa: E731

            # 调用 Triton 反向内核
            softmax_backward_kernel_inner[grid](
                out_reshaped,
                out_grad_reshaped,
                in_grad_reshaped,
                M * K,
                N,
                buffer_size_limit=2048,
            )
            # 将输入梯度恢复到原始布局
            # in_grad_view.copy_(in_grad_reshaped.view(M, K, N).transpose(1, 2))
            origin_dim = output.ndim
            if output.ndim == 3:
                m, n, k = output.shape
            elif output.ndim == 2:
                m, n = output.shape
            if M == 1 and origin_dim == 2:
                in_grad = in_grad_reshaped.view(K, N).transpose(0, 1)
            elif M == 1 and origin_dim == 3:
                in_grad = in_grad_reshaped.transpose(0, 1).view(m, n, k)
            else:
                in_grad = in_grad_reshaped.view(m, k, n).transpose(1, 2)
        else:
            grid = lambda meta: (M, 1, 1)  # noqa: E731

            softmax_backward_kernel_inner[grid](
                output,
                grad_output,
                in_grad,
                M,
                N,
                buffer_size_limit=2048,
            )
    return in_grad.to(input_dtype)
